src/core/ddsi/src/ddsi_handshake.c

```diff
@@ -553,7 +553,15 @@ static void func_validate_remote_identity(struct dds_security_fsm *fsm, void *ar
    * to be send.
    */
   if (handshake->local_auth_request_token.class_id && strlen(handshake->local_auth_request_token.class_id) != 0)
+  {
+    /* A short sleep to give the remote node some time for matching the
+       BuiltinParticipantStatelessMessageWriter, so that it won't drop the
+       sample we're sending (that would result in a time-out and the
+       handshake taking 1s longer than required) */
+    dds_sleepfor (DDS_MSECS(10));
```

Comment:
Ouch, an avoidable sleep in a thread that is supposed to do work on behalf of many independent entities. Adding a short delay for to cover this is a pragmatic solution, but surely there's a more elegant way to make this happen? Perhaps using an extra delay state in the FSM? That might be more complicated than it sounds, but it is worth exploring. If that proves to be too hard, then we can let this in as a temporary workaround with a comment.

Reply:
Yes, I agree that this is not elegant at all. Although blocking the FSM thread once per handshake won't have too much performance impact in many use cases I think, I will explore the option of adding the extra state in the FSM.